src/coreclr/src/jit/optimizer.cpp

```diff
@@ -8659,6 +8659,145 @@ bool Compiler::optIsRangeCheckRemovable(GenTree* tree)
+//      and convert them into single branchless expressions, e.g.
+//
+//            JTRUE
+//              \
+//              EQ              RETURN
+//             /  \                \
+//            X    Y               ADD
+//                       =>       /   \
+//           RETURN             EQ     C3
+//              \              /  \
+//              C1            X    Y
+//
+//           RETURN
+//              \
+//               C2
+//
+void Compiler::optBranchlessConditions()
+{
+    JITDUMP("\n*************** In optBranchlessConditions()\n");
+    bool changed = false;
+    for (BasicBlock* block = fgFirstBB; block; block = block->bbNext)
+    {
+        if ((block->bbJumpKind != BBJ_COND) || block->isEmpty())
+        {
+            continue;
+        }
+
+        Statement* fistStmt = block->lastStmt();
+        if ((fistStmt == nullptr) || !fistStmt->GetRootNode()->OperIs(GT_JTRUE))
+        {
+            continue;
+        }
+
+        GenTree*  rootNode    = fistStmt->GetRootNode();
+        GenTree*  rootSubNode = rootNode->gtGetOp1();
+        var_types typ         = rootSubNode->TypeGet();
+
+        assert(rootSubNode != nullptr);
+
+        if (!varTypeIsIntegral(typ) || (rootNode->gtFlags & GTF_ORDER_SIDEEFF))
+        {
+            continue;
+        }
+
+        BasicBlock* trueBb  = block->bbJumpDest;
+        BasicBlock* falseBb = block->bbNext;
+
+        assert((trueBb != nullptr) && (falseBb != nullptr));
+
+        if (!rootSubNode->OperIsCompare() || (trueBb == falseBb) ||
+            // both blocks are expected to be BBJ_RETURN
+            // TODO: implement for GT_ASG
+            (trueBb->bbJumpKind != BBJ_RETURN) || (falseBb->bbJumpKind != BBJ_RETURN) ||
+            // give up if these blocks are used by someone else
+            (trueBb->countOfInEdges() != 1) || (falseBb->countOfInEdges() != 1) ||
+            (trueBb->bbFlags & BBF_DONT_REMOVE) || (falseBb->bbFlags & BBF_DONT_REMOVE))
+        {
+            continue;
+        }
+
+        Statement* trueBbStmt  = trueBb->firstStmt();
+        Statement* falseBbStmt = falseBb->firstStmt();
+        if ((trueBbStmt == nullptr) || (trueBbStmt->GetPrevStmt() != trueBbStmt) ||
+            // make sure both blocks are single statement
+            (falseBbStmt == nullptr) || (falseBbStmt->GetPrevStmt() != falseBbStmt))
+        {
+            continue;
+        }
+
+        // make sure both blocks are single `return cns` nodes
+        GenTree* retTrueNode  = trueBbStmt->GetRootNode();
+        GenTree* retFalseNode = falseBbStmt->GetRootNode();
+        if (!retTrueNode->OperIs(GT_RETURN) || !retFalseNode->OperIs(GT_RETURN) ||
+            // both blocks of the same types
+            !retTrueNode->TypeIs(typ) || !retFalseNode->TypeIs(typ) ||
+            // both blocks are `return cns`
+            !retTrueNode->gtGetOp1()->IsIntegralConst() || !retFalseNode->gtGetOp1()->IsIntegralConst())
```

Comment:
>IsIntegralConst [](start = 38, length = 15)

So integral constant also include `long` on 32 bit target. Should we add some test cases to verify long that stores > 32-bit integer?

Reply:
Ah, looks like there is an issue on x86, looking at it now.